### ISSUE TITLE
UNWIND silently breaks query execution

### ISSUE THREAD

**Issue by sh-ghanique** [neo4j/neo4j#13209]
## UNWIND silently breaks query execution

- **Neo4j Version:** 5.8.0
- **Operating System:** Windows 11 (Probably irrelevant)
- **Relates to:** Importing JSON

If a JSON object contains an array, one would iterate over that array using an `UNWIND` statement. If the array is missing from the current data object, the query stops at that UNWIND statement without an error, and doesn't execute remaining statements.

### Working Example

Consider the following json file:

```json
{
    "name": "value",
    "tags": [
        "tag1",
        "tag2"
    ],
    "annotations": [
        "annotation1",
        "annotation2"
    ]
}
```

Consider the following cypher:

```cypher
WITH "file://path-to-json.json" AS uri
CALL apoc.load.json(uri) YIELD value AS json
MERGE (o:Object {name: json.name}

WITH o, json
UNWIND json.tags AS tag
MERGE (t:Tag {name: tag})
MERGE (o)-[:HAS_TAG]-(t)

WITH o, json
UNWIND json.annotations AS annotation
MERGE (o)-[:HAS_ANNOTATION]-(a:Annotation {name: annotation})
```

Importing the given JSON with the given Cypher results in:

- one Object node, 
- two HAS_TAG relationships with two Tag Nodes
- two HAS_ANNOTATION relationships with two Annotation Nodes

### Steps to reproduce

Consider the same Cypher query as above.

Consider the following JSON instead, where the tags are missing. 

```json
{
    "name": "value",
    "annotations": [
        "annotation1",
        "annotation2"
    ]
}
```

### Expected behavior

Either an Error
Or:

- one Object node, 
- two HAS_ANNOTATION relationships with two Annotation Nodes

### Actual behavior

- one Object node

Or, in other words, the part of the query where it iterates over the Annotations is never executed, without any indication that it is skipped, or why it is skipped.

**Comment by sh-ghanique** [comment]
Eventually I found the following workaround. I still believe it to be a bug, though.

```cypher
WITH "file://path-to-json.json" AS uri
CALL apoc.load.json(uri) YIELD value AS json
MERGE (o:Object {name: json.name}

WITH o, json
CALL {
    WITH o, json
    UNWIND json.tags AS tag
    MERGE (t:Tag {name: tag})
    MERGE (o)-[:HAS_TAG]-(t)
}

WITH o, json
CALL {
    WITH o, json
    UNWIND json.annotations AS annotation
    MERGE (o)-[:HAS_ANNOTATION]-(a:Annotation {name: annotation})
}
```

**Comment by InverseFalcon** [comment]
This isn't a bug, but I can try to explain the behavior and a workaround.

In Cypher, operations produce rows, and more importantly they execute per row. If you perform a MATCH that finds no matches, rows will go to zero, and as such subsequent operations will not be able to run, as there are no input rows to run upon.

UNWIND works similarly. UNWIND isn't really doing iteration. Instead, it's generating a cross product of the input row data with every element of the list, so you will get a row per list element. All of those rows are available for processing for subsequent operations, so it feels like an iteration from there, but it will produce zero rows when you run it on an empty or nonexistent list. When rows go to zero, your prior data is gone, there are no more records of data of which to execute upon.

You can run a PROFILE of the query to see this yourself. Watch the number of rows in the PROFILE plan. Most operations (aggregations and special operators aside) execute per incoming input row. Watch for when the rows go to zero. Operations after that won't have a row to execute upon.

You can recover from zero rows using an aggregation without a grouping key (like collect() or a count()), which should get you a row for that result. However, as mentioned, anything you had previously in the query is gone, it was wiped out with the row of data that held it.

Using a subquery (without a return) to address operations when performing something like this (a MATCH that may not exist, or an UNWIND over a possibly empty or nonexistent list) is definitely a good workaround here, as the subquery will execute for the input row, but will not change the cardinality, so the row you executed upon remains after the subquery finishes.

Using FOREACH instead can also work, though you are restricted to only write clauses within a FOREACH, so the subquery approach is far more flexible.

Another workaround is to use a CASE statement for a potentially missing or empty list. Something like:

```
UNWIND CASE WHEN size(coalesce(json.tags, [])) = 0 THEN [null] ELSE json.tags END as tag
```

An UNWIND of [null] will get you a single row where the variable is null, so it's a decent fallback to use, as it keeps the row and all its data, you can keep executing.

That said, having a more streamlined way to express this would be useful. I'm in favor of something like `OPTIONAL UNWIND ...` myself, but for now we don't have something like that in the language. Perhaps in the future.

**Comment by sh-ghanique** [comment]
Thank you for your elaborate response: for the explanation and for the other workarounds that could also work.

If I understand you correctly query execution ends when the number of rows becomes 0.

It makes sense to me that the statements following an UNWIND that produces 0 rows are not executed. **UNTIL the following WITH statement, that is**. After all, the second "WITH o, json" statement (and the UNWIND statement that follows) _does_ have rows.

The assertion that there are 0 rows (which stops the execution) is accurate within the context of the WITH statement, but it is wrong in the context of the entire query. Or, in other words, the execution engine incorrectly concludes that the remaining statements cannot (and therefore should not) be executed. That, IMHO, is the bug.

That said, with your explanation, https://neo4j.com/docs/cypher-manual/5/clauses/unwind/#unwind-using-unwind-with-an-empty-list indeed does say that it ceases query execution if the number of rows reaches zero. But even reading the documentation, it not executing remaining executable statements, is unexpected. This is probably because the RETURN statement includes the `empty` variable. I created https://github.com/neo4j/docs-cypher/pull/601 to make the illustration more clear.

Thanks again for your elaborate answer.

**Comment by InverseFalcon** [comment]
> It makes sense to me that the statements following an UNWIND that produces 0 rows are not executed. UNTIL the following WITH statement, that is. After all, the second "WITH o, json" statement (and the UNWIND statement that follows) does have rows.

Can you cite the part of the query you're talking about? If there is a subquery here in play that may explain it, but I want to make sure I'm looking at the same query (and part of that query) that you're looking at.

Also please include the PROFILE plan on this one as well, that may help for context.

**Comment by sh-ghanique** [comment]
What I mean is (see the comments in the query below):

```cypher
WITH "file://path-to-json.json" AS uri
CALL apoc.load.json(uri) YIELD value AS json
MERGE (o:Object {name: json.name}

WITH o, json
UNWIND json.tags AS tag // 0 rows
MERGE (t:Tag {name: tag}) // Nothing to do, 0 rows
MERGE (o)-[:HAS_TAG]-(t) // Nothing to do, 0 rows

WITH o, json // Hey, new things to do! 0 rows is no longer true!
UNWIND json.annotations AS annotation
MERGE (o)-[:HAS_ANNOTATION]-(a:Annotation {name: annotation})
```

The same can also be easily illustrated with the example from the documentation:

```cypher
UNWIND [] AS empty
RETURN "This value should be returned!"
```

IMHO the fact that it is documented that it doesn't return the string if the list is empty, doesn't make it less of a bug 😊. It could be a known issue, but I don't think it makes much sense to consider it "intended behavior", as illustrated with the comment "0 rows is no longer true!" in the query above.

For now, I know this is the way it is, and there are a few workarounds. So I'm good 😊. Thanks again for your elaborate response.

**Comment by InverseFalcon** [comment]
Honestly, the WITH clause showing rows again instead of staying at 0 seems like the bug to me, `o` and `json` should no longer be accessible since there are no rows for them to execute over. The only time we would expect rows to come back is if we're doing an aggregation.

Can you post the PROFILE plan of the query that shows this? We'd be interested in taking a look.

_____

And again, the not returning the string because rows go to zero is NOT a bug. It may not be intuitive, due to people misunderstanding how UNWIND and Cypher work, and certainly there is room for the documentation to improve here, but it's not a bug.

It's similar to how this query will not return a string:

```
MATCH (t:NonExistentLabel)
RETURN "Some return string"
```

When rows go to zero, then there is nothing for the next operation to execute upon. There are no rows to return, and so the string won't be returned.

**Comment by sh-ghanique** [comment]
I'm not sure I understand your remark. There clearly is more to do, both in my example as in the one in the documentation. Even though there are no tags, there are still annotations to be processed. The fact that it does add the annotations if you apply one of the workarounds is evidence of that.

I'm afraid I don't know how to produce the PROFILE, aside from the fact that I'm not near a computer at the moment. Both the JSON with the unexpected result and the Cypher query in question are in this issue, though.

Let's agree to disagree on whether this is a bug 😊 or not. Now I know this peculiarity exists, and there are workarounds, it's no longer relevant. For me it's no longer an issue.

I could agree with you on why the example from the documentation should not return a value. In a way my example is different, because it is not the main object that no longer has any rows, but a sub-object, if you may; the main object and other sub-objects still exist and do have rows.

There is a chance that one may only run into this issue when importing JSON objects with lists that can be absent if empty, like in the JSON in this issue.

**Comment by InverseFalcon** [comment]
You can PROFILE a query by prefixing the query with "PROFILE", and if running this through the Neo4j Browser, it will produce a visual execution plan, which will include the flow of rows between operations in the plan. 

https://neo4j.com/docs/cypher-manual/current/query-tuning/query-profile/

As mentioned, earlier, what you described sounded buggy to me, being able to go from 0 rows to 1 row without an aggregation or subquery, and a PROFILE plan (along with the accompanying query) might be helpful in determining if there is a bug in play or not.

**Comment by sh-ghanique** [comment]
I will try to see if I can get you the query plan later today.

It's not the same thing that goes to 0 rows. Or, not everything goes to 0 rows.

1. The main JSON object remains at 1 row.
2. The Tags turns out to be 0 rows, because the element is missing from this particular JSON file.
3. The Annotations would have turned out to be 2 rows, if query execution wouldn't stop prematurely.

1 has 1 row, 2 has 0 rows, 3 has 2 rows. 1 + 0 + 2 = 3 rows. The entire query only produces 1 row. 
1 was work, 2 doesn't, 3 has work.

Query execution ceases because it assumes that because there are no more rows for tags, that there are no more rows at all. In this scenario that assumption is wrong. And so it only produces the 1 row of 1. But there is more to do in this query than just 2. Despite of that there is no work in 2, work on 1 hasn't finished yet, and it should still start working on 3, which has 2 rows, which it doesn't.

(
1 is the first WITH statement that calls the apoc function that reads the JSON file.  
2 is the second WITH statement where it UNWINDs the json.tags.  
3 is the third WITH statement where it UNWINDs the json.annotations.  
)

**Comment by sh-ghanique** [comment]
Other perspectives, of which I realize they don't completely apply to this particular situation, are Atomicity, and Isolation.

Atomicity states that either the entire transaction is executed, or none of it is. As this JSON object has three rows (the root object with the name property, and two annotations), this transaction should either produce 0 rows or 3 rows, but it produces 1.

Isolation more or less says that, though transactions are executed in parallel, they behave as though they were executed in sequence, where the end state is the same whether the first transaction was executed first or last. If I apply this principle on this query (which is in a slightly different context than the context to which "Isolation" really applies), then it shouldn't make a difference if I first process the tags and then the annotations, or the other way around.

So

```cypher
WITH "file://path-to-json.json" AS uri
CALL apoc.load.json(uri) YIELD value AS json
MERGE (o:Object {name: json.name}

WITH o, json
UNWIND json.tags AS tag
MERGE (t:Tag {name: tag})
MERGE (o)-[:HAS_TAG]-(t)

WITH o, json
UNWIND json.annotations AS annotation
MERGE (o)-[:HAS_ANNOTATION]-(a:Annotation {name: annotation})
```

... and ...

```cypher
WITH "file://path-to-json.json" AS uri
CALL apoc.load.json(uri) YIELD value AS json
MERGE (o:Object {name: json.name}

WITH o, json
UNWIND json.annotations AS annotation
MERGE (o)-[:HAS_ANNOTATION]-(a:Annotation {name: annotation})

WITH o, json
UNWIND json.tags AS tag
MERGE (t:Tag {name: tag})
MERGE (o)-[:HAS_TAG]-(t)

```

... should be identical and produce the same output. But they do not! With the given JSON where this particular one doesn't have tags, the first produces 1 row and the second 3.

The shape of a tag is identical to that of an annotation: both are arrays of strings. So getting a different output by reversing the order of processing tags and annotations, is as unlikely (or, as unexpected) as

```cyher
SET a.name = "John"
SET a.age = 5
```

... and ...

```cyher
SET a.age = 5
SET a.name = "John"
```

... producing a different output. But it does produce a different output. To me that suggests that the ceasing of executing the rest of the query if an UNWIND's list is empty, is an odd behavior, to say the least.

**Comment by sh-ghanique** [comment]
Another way to describe the issue, is that when UNWIND decides to cease execution of the rest of the query, it only evaluates the variable it is asked to unwind. It's evaluating a smaller scope and applying it to a bigger scope. Instead execution should cease if all variables hold no rows anymore.

Yet another way to describe it, is that it looks like a garbage collector that releases memory for all variables if one function's variable falls out of scope because the function finished execution.

**Comment by sh-ghanique** [comment]
This is the query:

```cypher
PROFILE
WITH apoc.convert.fromJsonMap('
{
    "name": "value",
    "annotations": [
        "annotation1",
        "annotation2"
    ]
}
') AS json
MERGE (o:Object {name: json.name})

WITH o, json
UNWIND json.tags AS tag
MERGE (t:Tag {name: tag})
MERGE (o)-[:HAS_TAG]-(t)

WITH o, json
UNWIND json.annotations AS annotation
MERGE (o)-[:HAS_ANNOTATION]-(a:Annotation {name: annotation})
```

This is the result:

> Added 1 label, created 1 node, set 1 property, completed after 876 ms.

Here is the plan:

(The plan is interesting. It considers the annotations, but it falsely assumes there are no rows. After all, if you would reverse the order of processing the tags and annotations, then it would assume there are rows and create the Annotation nodes).

![plan](https://github.com/neo4j/neo4j/assets/108307713/b9cd5066-99b6-4a8f-a5db-2a7594ca6969)
![plan](https://github.com/neo4j/neo4j/assets/108307713/88081213-1d25-4e32-95e4-cc01dcc82a29)

**Comment by InverseFalcon** [comment]
Again, this is not a bug. The key thing to understand is that operations execute per row, variables are stored on each row, and if there is no row to execute upon, then it's no-op.

If you replaced the UNWIND and the merging of the tag and relationship with a MATCH on a nonexistent node, it would be the same result: for the given row, the MATCH would fail, the row would be wiped out, and subsequent operations would have nothing to execute upon.

It does not matter if the json object you're thinking about has valid properties or not; that json object is GONE along with the row, so `json.annotations` has no meaning, there is no json object to access data upon. The json variable and the data it had was on that row that was wiped out, it is no longer accessible.

________

There is no atomicity problem. Consider a simplified graph like this:

```
CREATE (:Start {name:'nonempty', endIds:[1,3,5]})
CREATE (:Start {name:'empty', endIds:[]})

CREATE (:End {id:5})
```

Now consider a query like this:

```
MATCH (start:Start {name:'nonempty'})
UNWIND start.endIds as endId
RETURN start, endId
```

We would get 3 rows, all with the same start node reference, and each row with a different endId, one of the elements of the list we unwound.

If we instead continued from this query instead of returning, and tried to match end nodes:

```
MATCH (start:Start {name:'nonempty'})
UNWIND start.endIds as endId
MATCH (end:End {id:endId})
CREATE (other:Other {id:endId})
RETURN start, end, other
```

We would get a single row back, consisting of the start node, the one :End node that matched, and only one newly created :Other node. 

This is because the MATCH executes for each of the 3 rows we had before. There are no end nodes with ids of 1 or 3, so those two rows get wiped out because the MATCH failed on those rows. 

The row where the endId = 5 succeeds, finding the end node with the same id.

That same row is the only row left to execute upon, so the CREATE only creates one :Other node, with an id of 5. 

In this example, the failure to create 3 :Other nodes is not a problem of atomicity. It isn't a problem at all. Operations can only execute per row, and if you filter out rows (either with a MATCH/WHERE that doesn't find any matches, or a WITH/WHERE to filter out rows, or UNWINDing a non-existent list or empty list), then subsequent operations have no row to execute upon.

_________

You would see something similar in one of your queries, if some of the rows in your json file has list elements, and some did not. The UNWINDs would wipe out the rows with nonexistent or empty lists, the same way those rows would be wiped out if you did a MATCH that failed to find anything based on values for those rows.

Likewise, once the rows are wiped out, either by a failing MATCH or an UNWIND on a missing or empty list, all subsequent operations that WOULD have been downstream would be no-op, as there are no rows to execute upon. That's what you're seeing in your case.

Think in terms of streams and filters. Anything filtered out is not going to make it downstream for subsequent operations. Anything that makes it through the filters will be able to execute those downstream operations.

I think you might be mistakenly thinking of clauses in a query as independent from each other, when they are not. Operations execute per row, and that means that filtering operations (including MATCHes that don't match, or UNWINDing of nonexistent or empty lists, which creates cross products with 0 list elements) will prevent later operations from executing upon those earlier rows, since those rows got wiped out.
_________

If you have operations that can wipe out the row, but you want to continue operations in any case, then you need to be either using language elements or techniques so that the row doesn't get wiped out.

OPTIONAL MATCH, for example, is a way to preserve the row even if the match fails. At a minimum, you would get a single row with a null for any newly introduced variables.

We don't have an OPTIONAL UNWIND yet, but the techniques referenced earlier are ways to get the same kind of behavior, making sure rows don't get wiped out.

Subqueries without a RETURN will keep the same cardinality as when they were called, so you can perform operations on them that would have wiped out the row, but outside the subquery the row persists.

FOREACH is an iterative processing clause (UNWIND is not, that is part of the confusion here) and doesn't change row cardinality, it can't wipe the row.

Hopefully this makes things a little clearer.

If you still have some confusion, I would advise you to try RETURNing at each line of the query and see what's on the rows returned, because those rows will be input to the next operation (operations execute per row).

**Comment by InverseFalcon** [comment]
Absolutely the order of your MATCHes makes a difference, especially if there are write operations in between your MATCHes, and one MATCH might result in 0 rows and the other won't.

```
MATCH (start:Start) // finds 2 nodes, so you get 2 rows, one for each
CREATE (:Thing) // creates 2 nodes, one per row
MATCH (:Nonexistent) // executed per row, but such nodes don't exist, all rows wiped out, rows go to 0
```

vs:

```
MATCH (:Nonexistent) // rows to go 0
CREATE (:Thing) // nothing created, no rows to execute upon
MATCH (start:Start) // nothing matched, no rows to execute MATCH upon
```

You would see the same reflected in a PROFILE plan of the query.

**Comment by InverseFalcon** [comment]
Also, your terminology is incorrect, with respect to transactions, atomicity, and isolation. 

The entire query executes in a single transaction; these are not multiple transactions executing in the same query. There are multiple CLAUSES that can execute in the same query or transaction (MATCH, MERGE, CREATE, etc), but there are no issues with isolation here (we're only talking about a single transaction, not multiple) or atomicity (the transaction itself executes in full or gets rolled back, but what will be a part of the commit at the end of the transactions depends upon what you're doing in the query; naturally, filtering out rows during processing will affect what will ultimately be written when the commit happens at the end of the query.

**Comment by sh-ghanique** [comment]
Thanks again for your elaborate response.

I think we both agree on why it happens. We may still disagree on whether it should happen 😊.

Perhaps my confusion lied in my perception of how WITH statements work, which I now refined based on your feedback: in no way it creates anything resembling a scope, which means that whatever follows the WITH statement can affect what is before the WITH statement (in this case, UNWINDing the tags removing the row that held the json variable). Obviously creating a sub query with CALL does, which prevents the row from being deleted.

Regardless, because I now know this behavior exists (and mostly understand it), I can probably work around it and be aware of other situations where this applies. So thanks again for your insight into the inner workings.

**Comment by InverseFalcon** [comment]
> I think we both agree on why it happens. We may still disagree on whether it should happen 😊.

You're free to critique the behavior and suggest a feature request or change (as noted, an OPTIONAL UNWIND would be a much more elegant way to handle the UNWIND issue). 

However, the behavior we're looking at right now is intended design, so in that sense, it isn't something we can consider a bug. I'm happy to continue to provide examples to clarify if there is any confusion.

> whatever follows the WITH statement can affect what is before the WITH statement

I'm not following you here, operations after the WITH should not affect what happened prior to the WITH clause, that would be VERY confusing, I can't imagine how that would allow for a sane language. Can you state your example here? You had multiple UNWINDs so I'm not sure which part of the query you're talking about.

Again, I think it may be helpful for comprehending what is happening if you see what is RETURNed after every row of Cypher, and if you analyze the PROFILE plan and pay attention to the number of rows that flow between the operators (rows that get filtered out before they reach an operator will not be executed by the operator). I can help clarify any questions you may have about the plan or Cypher behavior.

**Comment by InverseFalcon** [comment]
Maybe one part of this that might be giving you trouble is where the data is located?

When you reference something by a variable, such as `json.tags`, the `json` variable is part of the row data. That row data is the context for which the operation executes upon. That's why if you RETURN at certain points in your query, you will see the same JSON data referenced on each row, since each row has its own separate json reference (they just happen to be pointing at the same data). So each row is a record of data. Whether the row holds the data directly, or has a reference to a single common bit of data from earlier (such as the json object), the fact remains that whenever you reference any variable in Cypher, it is accessing it through the row that the operation is executing upon. 

When we perform a MATCH operation, that executes upon each row. When we apply a WHERE clause, it filters based on the data for each row. When we CREATE, we create a new node per row. When we RETURN, each row that remains is returned. And when we UNWIND a list, we generate a cross product of the row with every element of that list (which is why a cross product against 0 list elements wipes out the row). Any rows that get filtered out will not be available to be executed by the next operation in the plan.

Rows can be generated by operations (such as MATCH, UNWIND, MERGE, etc) and operations execute per row.

Hope that helps put things in perspective.

**Comment by InverseFalcon** [comment]
One other exercise for helping to understand UNWIND...

Remember UNWIND generates a cross product of the row with every element of the list you're UNWINDing.

If we start with this:

```
UNWIND [1,2,3] as one
RETURN one
```

We get 3 rows, with each number on its own row, for 1, 2, 3.

If we extend that query like this:

```
UNWIND [1, 2, 3] as one
UNWIND [4, 5, 6] as two
RETURN one, two
```

We will get a cross product of each of the rows from the first UNWIND (that we saw in the RETURN above) with each of the elements of the list, for a total of 9 rows:

```
one, two
1, 4
1, 5
1, 6
2, 4
2, 5
2, 6
3, 4
3, 5
3, 6
```

And if we extend it again, this time finishing with an UNWIND of an empty list, we would be left with no rows returned:

```
UNWIND [1, 2, 3] as one
UNWIND [4, 5, 6] as two
UNWIND [] as three
RETURN one, two, three
```

No rows returned, because the last UNWIND generates, per row, a cross product of that row against every element of the list (no elements).

But the work was still done. If we did a PROFILE of that query, we would see it go from 3 rows, to 9 rows, to 0 rows, so all the work still happens, just the last operation wipes out all the rows.

And if we change up the order, and have the UNWIND [] in the middle, we can see how rows go to zero and the last UNWIND has no rows to execute upon, leaving it as no-op:

```
UNWIND [1, 2, 3] as one
UNWIND [] as two
UNWIND [4, 5, 6] as three
RETURN one, two, three
```

PROFILE that, and you'll see rows go from the initial 3 to 0, and then stay at zero. The last UNWIND has no rows to execute upon, so the operation never gets called. Likewise, the RETURN doesn't have any rows to return.

**Comment by sh-ghanique** [comment]
Yeah, that's as much as I figured out from your previous explanations 😊. Your example illustrates it really well.

To answer your question about my remarks about statements after the WITH statement affecting statements before it:

If ...

```cypher
MERGE (o:Object {name: json.name})
WITH json, o
UNWIND json.tags AS tag
```

... results in every `json` and `o` being cross joined with every `tag` (which results in 0 rows because there are 0 tags), then the statement after the WITH statement (the UNWIND) affects the statement before it (the MERGE statement), because there used to be an `o` before the WITH statement, and after the UNWIND there no longer is.

**Comment by InverseFalcon** [comment]
> the statement after the WITH statement (the UNWIND) affects the statement before it (the MERGE statement), because there used to be an o before the WITH statement, and after the UNWIND there no longer is.

That's not quite true, I think you may be confusing the presence of a the `o` variable in the row data (which IS gone because the UNWIND of the empty list wiped out the rows) and the presence of a new :Object node in the graph with the given name, which there IS because the MERGE executed, and so the node will be a part of the committed data for this transaction.

So there is a difference between the presence and accessibility of a variable reference, and the presence of the node in the graph as a result of that MERGE.

So if you ran this query, and then afterward ran a new query with `MATCH (n:Object {name: 'value'}) RETURN n` you would see that the node had been properly created in the earlier query.

Same thing if you had removed the `o` from scope in your WITH clause:

```
...
MERGE (o:Object {name: json.name})
WITH json
...
```

Because the `o` isn't present in the WITH clause, `o` would be dropped of scope and thus unavailable to use later in the query, but the MERGE had already executed, so provided the query didn't error out and rollback, the new :Object node would have been added into the graph when the transaction commits at the end.

Another similar example:

```
CREATE (o:Thing {id:123})
WITH o
WHERE false
RETURN 'some string'
```

Even though all rows get filtered out and nothing is returned, the CREATE was still successful and the node is committed into the graph.

So if you execute a write clause of any sort, even if rows go to zero afterward, as long as the transaction succeeds the changes will be committed into the graph.

**Comment by sh-ghanique** [comment]
I'm not confusing the two 😊. I know the Object node is in the graph, and I understand now how the o variable no longer contains any rows after the UNWIND.

What I meant to say was that the UNWIND after the WITH statement affected the value / content of the o variable, which was assigned a value (in this case 1 row) before the WITH statement.

But it's probably not important 😊.

**Comment by InverseFalcon** [comment]
> I know the Object node is in the graph, and I understand now how the o variable no longer contains any rows after the UNWIND.

Close...more like there is no row to contain any variables, so the `o` variable isn't accessible because there is no row to access it through or execute upon.

>What I meant to say was that the UNWIND after the WITH statement affected the value / content of the o variable, which was assigned a value (in this case 1 row) before the WITH statement.

I still think there's some confusion here.

Variables only exist for the duration of a query, they are temporary things and have no meaning outside of it. Once the query is done `o` doesn't have any meaning, but that would ALWAYS be the case no matter if no rows were filtered out during its execution.

But what gets created or changed or deleted in a query, provided the transaction doesn't roll back, gets committed into the graph.

_________

There isn't any kind of retroactive behavior here.

The MERGE executed, so that will either match to an existing node or create a new one, and that will be staged to commit when the transaction finishes. That new node is assigned to the `o` variable for that row.

Next the UNWIND happens, and the cross product x 0 elements filters the row. The row is also what held the `o` and `json` variables, so without a row to execute upon, all subsequent operations will be no-op (or more accurately, that row that had that data will not be executed by subsequent operations because it is now gone; any other rows that DID have a non-empty list would continue executing as normal, since they weren't filtered out).

So there's nothing retroactive happening, an UNWIND of an empty list is like filtering out the row, you'd get the same behavior if you did `WITH json, o WHERE size(coalesce(json.tags, [])) > 0 ...`

**Comment by sh-ghanique** [comment]
> Close. There are no more rows that contain variables.

Oh. Right. Of course. "Rows having variables" has quite different implications from "variables having rows".

Let me re-read everything you said. This may very well complete the picture, and explain why, indeed, it makes sense that execution ceases after an UNWIND against an empty list, where whether or not there were previous WITH statements is irrelevant.

**Comment by sh-ghanique** [comment]
I think I got it now.

It's funny. I've worked with Neo4j on and off for a number of years now, and fell in love with how a Graph allows one to gather data from multiple domains and to interconnect them, and how it allows one to think (read: Query) about data differently. Yet never did I really get the concept of rows as you explained in this thread, how rows have variables, and how they are not just something that is used in the output of a query to logically separate one return value from another.

Knowing what I know now I scanned the documentation (https://neo4j.com/docs/cypher-manual/5/introduction/) asking myself where did I go wring? Should I have known if I would have read the documentation more carefully? I'm not sure. Maybe from the Execution Plans section.

I remember that a long time ago didn't see the point of the WITH statement, except for that you needed it when some queries resulted in an error that stated that you did need to use it. Eventually I read the respective page: https://neo4j.com/docs/cypher-manual/5/clauses/with/. Perhaps that is what threw me off.

The page stars with introducing "Query Parts", and in the next block it talks about "Scopes" and that "Any variables not included in the WITH clause are not carried over to the rest of the query". Maybe that is how I came to the conclusion that queries are executed by Variables (not by rows), and that when you use a WITH statement, that you, in a way, make a "copy" of those variables and put them on the Stack, if you may. And when one WITH statement completes, you take the same _values_ as before, and feed them to the next WITH statement, where they have the same values as before the previous WITH statement... In the picture that I had in my mind the query was executed "in branches" (where every WITH statement creates a new branch), similar to a recursive function in basically any programming language.

Though, after this conversation, this mental model turns out to be incorrect, applying it has always worked out well until now. I guess that is why it took you so much effort to change it, and why it took me so much effort to understand what you were saying 😊. (Lots of respect for your patience!)

Now I understand that the query doesn't execute in branches, but that every statement (but also every query section) makes mutations to the query's "Shared State", which is a collection of Rows. Those mutations can include changes to the content of each Row, but also to the collection of Rows.

So to conclude:

- I can't thank you enough for your patience, and for the new understanding.
- It is NOT a bug (because UNWIND doesn't branch off, but modifies the collection of rows).
- An OPTIONAL UNWIND could be a great addition (I'm not going to chase it, though).

If it's not just me, then maybe the documentation could use another page that explains how the engine executes queries, how certain clauses affect the collection of Rows it's working on, and how variables fit in this picture. But I leave that up to you 😊.

Thanks again!

I really appreciate your effort.

**Comment by InverseFalcon** [comment]
Always good to reach a better understanding, and glad to help! Reminds me of when I was learning Cypher myself, UNWIND was a tripping point for me as well. The "operators produce rows, and operators execute per row" part was one of the key realizations for me back then, and it's something I like to lead with as part of the foundation for understanding Cypher. 

There's always room for improvement on documentation and teaching, I'll see if we can look through and use anything here to improve what we have.
